packages/server/src/queries/complex/portfolio/allocation.ts

```diff
@@ -0,0 +1,184 @@
+      amount: unclaimedRewardsCap,
+    },
+    {
+      key: "pooled",
+      percentage: (pooledCap / totalCap) * 100,
+      amount: pooledCap,
+    },
+  ];
+}
+
+async function getAssets(categories: Categories, assetLists: AssetList[]) {
+  const totalAssets = categories["total-assets"];
+  const totalCap = +totalAssets.capitalization;
+
+  // Get top 5 assets by cap value
+  const sortedAccountCoinsResults =
+    totalAssets?.account_coins_result?.sort(
+      (a: AccountCoinsResult, b: AccountCoinsResult) =>
+        +b.cap_value - +a.cap_value
+    ) || [];
+
+  const top5AccountCoinsResults = sortedAccountCoinsResults.slice(0, 5);
+
+  const assets = top5AccountCoinsResults.map((asset: AccountCoinsResult) => {
+    const assetFromAssetLists = getAsset({
+      assetLists,
+      anyDenom: asset.coin.denom,
+    });
+
+    return {
+      key: assetFromAssetLists.coinDenom,
+      percentage: (+asset.cap_value / totalCap) * 100,
+      amount: +asset.cap_value,
+    };
+  });
+
+  const otherAssets = sortedAccountCoinsResults.slice(5);
+  const otherAmount = otherAssets.reduce(
+    (sum: number, asset: AccountCoinsResult) => sum + +asset.cap_value,
+    0
+  );
+  const otherPercentage = (otherAmount / totalCap) * 100;
+
+  const other = {
+    key: "Other",
+    percentage: otherPercentage,
+    amount: otherAmount,
+    color: "bg-osmoverse-500",
+  };
+
+  return [...assets, other];
+}
+
+async function getAvailable(categories: Categories, assetLists: AssetList[]) {
+  const userBalances = categories["user-balances"];
+  const totalCap = +userBalances.capitalization;
+
+  // Get top 5 assets by cap value
+  const sortedAccountCoinsResults =
+    userBalances?.account_coins_result?.sort(
+      (a: AccountCoinsResult, b: AccountCoinsResult) =>
+        +b.cap_value - +a.cap_value
+    ) || [];
+
+  const top5AccountCoinsResults = sortedAccountCoinsResults.slice(0, 5);
+
+  const assets = top5AccountCoinsResults.map((asset: AccountCoinsResult) => {
+    const assetFromAssetLists = getAsset({
+      assetLists,
+      anyDenom: asset.coin.denom,
+    });
+
+    return {
+      key: assetFromAssetLists.coinDenom,
+      percentage: (+asset.cap_value / totalCap) * 100,
+      amount: +asset.cap_value,
+    };
+  });
+
+  const otherAssets = sortedAccountCoinsResults.slice(5);
+  const otherAmount = otherAssets.reduce(
+    (sum: number, asset: AccountCoinsResult) => sum + +asset.cap_value,
+    0
+  );
+  const otherPercentage = (otherAmount / totalCap) * 100;
+
+  const other = {
+    key: "Other",
+    percentage: otherPercentage,
+    amount: otherAmount,
+  };
+
+  return [...assets, other];
+}
+
+export async function getAllocation({
+  address,
+  assetLists,
+}: {
+  address: string;
+  assetLists: AssetList[];
+}): Promise<GetAllocationResponse> {
+  return await cachified({
+    cache: allocationCache,
+    ttl: 1000 * 60 * 0.25, // 15 seconds since a user can transact quickly
```

Comment:
I'd avoid adding caching at all to this query. A main reason it was added to SQS was that it contained all user data and needed the cache miss performance gains. I think it would be a good idea to completely avoid the risk of old balances appearing after the queries are invalidated in the client after a transaction.

Then, they'd only later update if they become stale on the client, which could take a while.